Reconfigure container manifests to install using RPMs

kargakis opened the pull request:
Picks up https://github.com/openshift/origin/pull/12510

smarterclayton commented on Makefile:
```diff
@@ -286,13 +286,14 @@ build-rpms-redistributable:
 #   make release-rpms
-release-rpms: clean build-rpms
+release-rpms: clean build-rpms-redistributable
+	hack/build-base-images.sh
```
Why is this here?

smarterclayton replied:
Because we put things in it now?

kargakis replied:
Likely but @stevekuznetsov can confirm

stevekuznetsov replied:
> Because we put things in it now?

What does this mean?

stevekuznetsov replied:
Oh -- the flow now is build RPMs, build base images with mounted RPMs, etc -- we do need RPMs to be present before the image builds

smarterclayton replied:
Why do we have `clean` here?

stevekuznetsov replied:
Copied from `make release`

kargakis replied:
> Oh -- the flow now is build RPMs, build base images with mounted RPMs, etc -- we do need RPMs to be present before the image builds

`make release` is also building images and now we'll require the RPMs to be present. Should we also change that?

kargakis replied:
Or we could have a target that builds the RPMs and another one that builds the images. Right now, most targets overlap.

stevekuznetsov replied:
If @smarterclayton is fine with it, this should replace the previous `make release`

kargakis replied:
> Right now, most targets overlap.

s/most/release vs release-rpms/

> If @smarterclayton is fine with it, this should replace the previous make release

+1

smarterclayton commented on hack/common.sh:
```diff
@@ -68,17 +68,9 @@ readonly -f os::build::platform_arch
-  fi
-
-  if [[ -z "$(which sha256sum)" ]]; then
+  if os::util::find::system_binary 'sha256sum' >/dev/null 2>&1; then
```
This is not what the previous code was doing.

kargakis replied:
good catch

smarterclayton commented on hack/common.sh:
```diff
@@ -91,13 +83,8 @@ EOF
-
-EOF
-      exit 2
+      os::log::fatal "Detected Go version: ${go_version[*]}.
```
Test doesn't match message.

kargakis replied:
Which test? Does this need the extra empty lines?

smarterclayton replied:
Test is for a different version than the message

kargakis replied:
fixed

smarterclayton commented on hack/lib/build/release.sh:
```diff
@@ -0,0 +1,14 @@
+
+# os::build::release::check_for_rpms checks that an RPM release has been built
+function os::build::release::check_for_rpms() {
+	if [[ ! -d "${OS_LOCAL_RPMPATH}" || ! -s "${OS_LOCAL_RELEASEPATH}/CHECKSUM" ]]; then
```
Why do you have to check for something in the release path?

kargakis replied:
It seems that the RPMs are in the release path.

stevekuznetsov replied:
RPMs are at `_output/local/releases/rpms`

kargakis replied:
@stevekuznetsov now I think the `! -s "${OS_LOCAL_RELEASEPATH}/CHECKSUM` is unecessary here since the release binaries are not being built as part of building the RPMs and the images have no use for the release binaries anyway.

kargakis replied:
>  and the images have no use for the release binaries anyway.

Or at least they don't have any use since now we switch to placing the RPMs inside the images (I think only the origin-pod and dockerregistry images required the existence of a binary prior to this PR).

stevekuznetsov replied:
> release binaries are not being built as part of building the RPMs

`rpmbuild` calls `hack/build-cross.sh` -- shouldn't this result in binaries, same as the old release process?

stevekuznetsov replied:
In my investigation before I had placed a `set -o xtrace` into the build and I saw the CHECKSUM etc get placed down, but some of these files for whatever reason were not getting copied back from the `hack/env` container AFAIR

kargakis replied:
Yeah, now I see. For some reason I am not getting back from `hack/env` anything related to the release binaries.. Looking further..

kargakis replied:
So it seems we just need to preserve everything in the release directory if we want CHECKSUM back:) Updated the suggestion.

smarterclayton commented on images/base/Dockerfile:
```diff
@@ -1,17 +1,17 @@
     mkdir -p /var/lib/origin
 
-LABEL io.k8s.display-name="OpenShift Origin Centos 7 Base" \
+LABEL io.openshift.tags="openshift,base" \
```
Why does this image need tags?

kargakis replied:
@stevekuznetsov

stevekuznetsov replied:
I think this came from what we have in OSBS right now

smarterclayton replied:
Why in the world would these be important?

stevekuznetsov replied:
Not my fight to have ... @tdawson

tdawson replied:
LABEL io.openshift.tags is supposed to help with the Container Catalog.
https://access.redhat.com/containers/#/
But, for this image, I have to agree, this isn't important.  This image isn't pushed to the Container Catalog.
I know I was the one who put it in, but I'm fine with taking it out for this image.

smarterclayton replied:
@tdawson did anyone review the tags for catalog impact, or just basic ones?

We can leave in place for now but would like to follow up.

smarterclayton commented on images/deployer/Dockerfile:
```diff
@@ -12,8 +12,9 @@
 FROM openshift/origin
 
-LABEL io.k8s.display-name="OpenShift Origin Deployer" \
+LABEL io.openshift.tags="openshift,deployer" \
```
Doesn't need tags

stevekuznetsov replied:
Ditto, from OSBS

smarterclayton commented on images/dockerregistry/Dockerfile:
```diff
@@ -6,16 +6,20 @@
-      io.k8s.description="This is a component of OpenShift Origin and exposes a Docker registry that is integrated with the cluster for authentication and management."
+COPY config.yml ${REGISTRY_CONFIGURATION_PATH}
+
+LABEL io.openshift.tags="openshift,docker,registry" \
```
Doesn't need tags

stevekuznetsov replied:
Ditto, from OSBS

smarterclayton commented on images/dockerregistry/Dockerfile:
```diff
@@ -6,16 +6,20 @@
 
-LABEL io.k8s.display-name="OpenShift Origin Image Registry" \
-      io.k8s.description="This is a component of OpenShift Origin and exposes a Docker registry that is integrated with the cluster for authentication and management."
+COPY config.yml ${REGISTRY_CONFIGURATION_PATH}
```
why was this reordered?

kargakis replied:
@stevekuznetsov

stevekuznetsov replied:
Unintentional AFAICT

smarterclayton commented on images/dockerregistry/Dockerfile:
```diff
@@ -6,16 +6,20 @@
 ENV REGISTRY_CONFIGURATION_PATH=/config.yml
 
-CMD DOCKER_REGISTRY_URL=${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT} /dockerregistry ${REGISTRY_CONFIGURATION_PATH}
+CMD DOCKER_REGISTRY_URL=${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT} /usr/bin/dockerregistry ${REGISTRY_CONFIGURATION_PATH}
```
We shouldn't be setting ENV vars like this in the command - we should fix that now.

kargakis replied:
So if I move this as an ENV, `${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT}` resolves to `:` whereas in `CMD` it resolves to an empty string. I am not sure why `CMD` does that - I have opened https://github.com/openshift/origin/pull/14070 where I am looking into getting rid of the ENV altogether since it's only used for testing.

smarterclayton commented on images/node/Dockerfile:
```diff
@@ -8,28 +8,23 @@
-COPY opt/cni/bin/* /opt/cni/bin/
-COPY etc/cni/net.d/* /etc/cni/net.d/
-COPY conf/openshift-sdn-ovs.conf /usr/lib/systemd/system/origin-node.service.d/
 COPY scripts/* /usr/local/bin/
```
Why aren't these scripts in the RPM?

smarterclayton commented on images/node/Dockerfile:
```diff
@@ -8,28 +8,23 @@
-COPY etc/cni/net.d/* /etc/cni/net.d/
-COPY conf/openshift-sdn-ovs.conf /usr/lib/systemd/system/origin-node.service.d/
 COPY scripts/* /usr/local/bin/
 COPY system-container/system-container-wrapper.sh /usr/local/bin/
```
Should be in the RPM

smarterclayton commented on images/node/Dockerfile:
```diff
@@ -8,28 +8,23 @@
+    mkdir -p /usr/lib/systemd/system/docker.service.d /usr/lib/systemd/system/origin-node.service.d
 
-LABEL io.k8s.display-name="OpenShift Origin Node" \
+LABEL io.openshift.tags="openshift,node" \
```
No tags

stevekuznetsov replied:
If we want/need them in OSBS builds.... are you going to live with them here?

smarterclayton commented on images/openvswitch/Dockerfile:
```diff
@@ -8,10 +8,15 @@ FROM openshift/node
 COPY scripts/* /usr/local/bin/
```
Should be in one of the RPMs

smarterclayton commented on images/origin/Dockerfile:
```diff
@@ -12,24 +12,20 @@ FROM openshift/origin-base
+    yum install -y ${INSTALL_PKGS} && \
+    rpm -V ${INSTALL_PKGS} && \
+    yum clean all && \
     setcap 'cap_net_bind_service=ep' /usr/bin/openshift
```
Should be done by the RPM

kargakis replied:
Requires that rpmbuild runs as root

stevekuznetsov replied:
Isn't that true for Brew? /cc @tdawson and if it is, we should make it true for Origin

tdawson replied:
I don't see how this makes rpmbuild need to be run as root.
It is not true that Brew builds run as root, just the opposite.  It screams at you if you try to do anything that requires root priviledges.

kargakis replied:
Which means that setcap needs to stay in the Dockerfile

smarterclayton commented on images/dockerregistry/Dockerfile:
```diff
@@ -6,16 +6,20 @@
+
+LABEL io.openshift.tags="openshift,docker,registry" \
+      io.k8s.display-name="OpenShift Container Platform Image Registry" \
+      io.k8s.description="This is a component ofOpenShift Container Platform and exposes a Docker registry that is integrated with the cluster for authentication and management."
```
Typo, also, these are origin images, so "Container Platform" isn't really appropriate. 

We should use "OpenShift" everywhere in these.

kargakis commented on origin.spec:
```diff
@@ -244,6 +244,7 @@ do
   echo "+++ INSTALLING ${bin}"
   install -p -m 755 _output/local/bin/${PLATFORM}/${bin} %{buildroot}%{_bindir}/${bin}
 done
+# setcap 'cap_net_bind_service=ep' %{buildroot}%{_bindir}/openshift
```
Not sure where to place setcap inside the RPM spec and it also requires root.

cc: @smarterclayton @tdawson @stevekuznetsov

kargakis commented on images/pod/Dockerfile:
```diff
@@ -1,16 +1,22 @@
 # The standard name for this image is openshift/origin-pod
 #
-FROM scratch
+FROM openshift/origin-source
```
I guess this change is going to grow the origin-pod image by a fair amount since origin-source is based on centos. @smarterclayton @stevekuznetsov thoughts?

kargakis replied:
It seems that this has already changed for OSBS images (they are using rhel7.3). Doesn't this mean that every pod will run with the additional size of a rhel image on top of the application image? :)

kargakis replied:
# Pulp can't handle the scratch later currently

hmm..

stevekuznetsov replied:
We pay the cost once per node, right?

kargakis replied:
Not sure, is this per node or per pod @smarterclayton ?

stevekuznetsov replied:
Would want resolution here @smarterclayton @tdawson

stevekuznetsov commented on hack/common.sh:
```diff
@@ -84,7 +84,7 @@ function os::build::setup_env() {
     if [[ "${go_version[2]}" < "go1.5" ]]; then
       os::log::fatal "Detected Go version: ${go_version[*]}.
-Origin builds require Go version 1.6 or greater."
+Origin builds require Go version 1.5 or greater."
```
This matches the check above but don't we want 1.7 right now?

kargakis replied:
Beats me what we want

stevekuznetsov commented on images/base/Dockerfile:
```diff
@@ -13,5 +13,5 @@ RUN INSTALL_PKGS="bsdtar ceph-common device-mapper device-mapper-persistent-data
     mkdir -p /var/lib/origin
 
 LABEL io.openshift.tags="openshift,base" \
-      io.k8s.display-name="OpenShift Origin CentOS 7 Base" \
```
Can we do this?

kargakis replied:
Based on https://github.com/openshift/origin/pull/13642#discussion_r110762845 I figured we can

kargakis replied:
Reverted

stevekuznetsov commented on hack/common.sh:
```diff
@@ -70,7 +70,7 @@ readonly -f os::build::platform_arch
   os::util::ensure::system_binary_exists 'go'
 
-  if os::util::find::system_binary 'sha256sum' >/dev/null 2>&1; then
+  if [[ -z "$(which sha256sum)" ]]; then
```
I guess this is technically not the same as `os::util::find::system_binary` .... but that was the _intent_ behind this check, right? If we don't have the program, declare a no-op. Re-declaring when we've already declared is fine.

stevekuznetsov commented on images/source/Dockerfile:
```diff
@@ -8,4 +8,4 @@
 
 ADD origin-local-release.repo /etc/yum.repos.d/
-ADD centos-paas-sig-openshift-future.repo /etc/yum.repos.d/
+ADD centos-paas-sig-openshift-origin36.repo /etc/yum.repos.d/
```
I don't really understand this commit -- is this just to bring in stuff without rebasing?

kargakis replied:
There is no need for rebasing since the existing origin36 changes in master are in the node image. I figured that it should change for the source image, is this not what we want?

stevekuznetsov replied:
Oh, I entirely misread. Yes I think this is correct.

kargakis replied:
So adding this to the source images makes the base image confused about which yum repo to choose. I think we need to handle this on a case-by-case basis. Also I am not sure we need source.

stevekuznetsov commented on images/node/Dockerfile:
```diff
@@ -10,8 +10,6 @@
 FROM openshift/origin
 
-COPY scripts/* /usr/local/bin/
-COPY system-container/system-container-wrapper.sh /usr/local/bin/
```
Are we certain that we want the system container stuff in the RPMs? I thought we were going to ship that in some other manner /cc @tdawson

tdawson commented:
I really like the Dockerfile cleanup.  Having a more generic description will really help with the differences between Origin and OCP.
The only concern I have is with the changes in hack/common.sh.  I seem to recall something similar being tried before.  Anyway, I'm going to give this my usual manual test and see how it goes.

kargakis commented:
> The only concern I have is with the changes in hack/common.sh. I seem to recall something similar being tried before. Anyway, I'm going to give this my usual manual test and see how it goes.

@tdawson this is not ready yet. I'll ping you when it is.

kargakis commented:
[test]

kargakis commented:
@stevekuznetsov got a successful run of `hack/env make release` with what I have here, trying to see if everything works fine with `OS_BUILD_ENV_PRESERVE=_output/local/bin:_output/local/releases hack/env make release`.

kargakis commented:
> trying to see if everything works fine with OS_BUILD_ENV_PRESERVE=_output/local/bin:_output/local/releases hack/env make release.

Also works, I am getting back all files from the container.

stevekuznetsov commented:
Awesome. @smarterclayton @tdawson final reviews?

tdawson commented:
Looked good.
I did manual test and it passed.
👍

stevekuznetsov commented on Makefile:
```diff
@@ -236,8 +236,9 @@ clean:
-	hack/build-images.sh
+	OS_ONLY_BUILD_PLATFORMS="linux/amd64" hack/build-rpm-release.sh
+	hack/build-base-images.sh
+	OS_BUILD_ENV_EXCLUDE='_output/local/pkgdir' hack/build-images.sh
```
`hack/env` stuff should be applied outside of `make`, not inside of it

stevekuznetsov commented on Makefile:
```diff
@@ -236,8 +236,9 @@ clean:
 release: clean
-	OS_ONLY_BUILD_PLATFORMS="linux/amd64" hack/build-release.sh
-	hack/build-images.sh
+	OS_ONLY_BUILD_PLATFORMS="linux/amd64" hack/build-rpm-release.sh
```
list `build-rpms` as a child of this target, don't call the script yourself

stevekuznetsov commented on hack/build-images.sh:
```diff
@@ -8,46 +8,14 @@
 source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
-source "${OS_ROOT}/contrib/node/install-sdn.sh"
-
-if [[ "${OS_RELEASE:-}" == "n" ]]; then
```
We need to be clear to DevEx that the UX for this is changing, e-mail should be fine (along with the good error message in here if RPMs have not yet been built)

stevekuznetsov commented on hack/build-images.sh:
```diff
@@ -8,46 +8,14 @@
-	os::build::extract_tar "${OS_IMAGE_RELEASE_TAR}" "${imagedir}"
-fi
+# we need to mount RPMs into the container builds for installation
+OS_BUILD_IMAGE_ARGS="${OS_BUILD_IMAGE_ARGS:-} -mount ${OS_LOCAL_RPMPATH}/:/srv/origin-local-release/"
```
This script needs to depend on `imagebuilder` too

kargakis replied:
Done

stevekuznetsov commented on hack/build-rpm-release.sh:
```diff
@@ -3,9 +3,10 @@
+# consume all of the relevant envars.
 source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
+os::util::ensure::system_binary_exists tito
+os::util::ensure::system_binary_exists createrepo
```
Why the change?

kargakis replied:
If `createrepo` does not exist, we will not fail, the repodata file for the yum repo that we are mounting to the images will not be created, and eventually our images aare broken.

stevekuznetsov replied:
Oh, right, we need to make the repo now to mount it. Ack.

stevekuznetsov commented on hack/lib/log/output.sh:
```diff
@@ -27,13 +27,6 @@ function os::log::warning() {
 readonly -f os::log::warning
 
-# os::log::warn remains for backward compatibility
-function os::log::warn() {
```
When squashing let's keep this as a separate commit

stevekuznetsov commented on images/base/Dockerfile:
```diff
@@ -1,17 +1,17 @@
-LABEL io.k8s.display-name="OpenShift Origin Centos 7 Base" \
-      io.k8s.description="This is the base image from which all OpenShift Origin images inherit."
+LABEL io.openshift.tags="openshift,base" \
+      io.k8s.display-name="OpenShift CentOS 7 Base" \
```
Did we get the sign-off for less branding here?

kargakis replied:
restored to use Origin

stevekuznetsov commented on images/node/Dockerfile:
```diff
@@ -10,13 +10,6 @@
-COPY conf/openshift-sdn-ovs.conf /usr/lib/systemd/system/origin-node.service.d/
-COPY scripts/* /usr/local/bin/
-COPY system-container/system-container-wrapper.sh /usr/local/bin/
 COPY system-container/manifest.json system-container/config.json.template system-container/service.template system-container/tmpfiles.template /exports/
```
Why did these stay but the `system-container-wrapper.sh` go?

kargakis replied:
Mainly because of the naming but I could rename them similarly to how I renamed the scripts so they won't collide.

stevekuznetsov replied:
@giuseppe are the names here important enough? If not, let's skip the rename and put them in the RPMs

kargakis replied:
Unless I move each one (`openvswitch`, `node`, `origin`) in its own directory, they will collide in the RPM, right?

stevekuznetsov replied:
They should be in different RPMs, right? Not sure about OVS (also not clear why our repository is where the OVS systemd files are) but the other could just be in `origin` and `origin-node`

giuseppe replied:
@stevekuznetsov yes, the name are important and they must be under /exports/ as well.

I think they won't make much sense in the .rpm itself (also you probably don't want to create the non standard path /exports).  I'd leave them in the Dockerfile if that is possible.

stevekuznetsov replied:
@smarterclayton @giuseppe let's determine for certain whether we do or do not want these files in the RPM. Also, what the generally accepted manner is to support system containers. I know @tdawson has been asked to set up new system container image creations in Brew, right? If we're doing those builds separately, do we need these files in the Origin container images?

kargakis replied:
Does this discussion need to happen in this PR? I can revert the changes and move the system container stuff back in the Dockerfiles and we can move them back to RPMs later if we decide to @smarterclayton ?

stevekuznetsov replied:
No, I don't think so. That is why I did not originally tackle it in my PR.

kargakis replied:
restored the system container stuff in the dockerfiles

stevekuznetsov commented on images/pod/Dockerfile.centos7:
```diff
@@ -22,6 +22,6 @@ RUN INSTALL_PKGS="origin-pod" && \
     ln -s /usr/bin/pod /pod
```
In what cases do we want to carry diffs between `Dockerfile` and `Dockerfile.centos7` ?

kargakis replied:
Not sure. A question for @tdawson @smarterclayton

tdawson replied:
That line is/was only in the Dockefile.centos7 because the ENTRYPOINT is/was "/pod".  If the ENTRYPOINT in the normal Dockerfile is changed, then it should be changed in Dockerfile.centos7, and the "ln -s /usr/bin/pod /pod" line should be taken out of Dockerfile.centos7

kargakis replied:
Will make sure all Dockerfiles and Dockerfiles.centos7 are synced

stevekuznetsov commented on origin.spec:
```diff
@@ -311,6 +311,18 @@ install -m 0644 contrib/tuned/man/tuned-profiles-origin-node.7 %{buildroot}%{_ma
+install -d %{buildroot}/usr/local/bin/
+
+# Install master scripts
+install -p -m 0755 images/origin/system-container/origin-system-container-wrapper.sh %{buildroot}/usr/local/bin/
```
@giuseppe and @tdawson would want general confirmation that the changes to system container stuff here is kosher and that we are OK to be putting this into the RPMs

tdawson replied:
If we are putting it into an rpm, it should be in /usr/bin/ not /usr/local/bin/

kargakis replied:
I guess I can change the rest of the scripts but for the `node` image there is an issue with /usr/bin and docker: https://github.com/openshift/origin/blob/master/images/node/scripts/docker#L14

kargakis commented:
In short, I've reverted the changes for the system container scripts, reverted branding, and I've added a last commit that tries to sync Origin with the images in dist-git as much as possible.  PTAL

kargakis commented:
@stevekuznetsov I think we need to update some jobs that are calling build-images.sh w/o building RPMs..

@smarterclayton ptal

stevekuznetsov commented:
Yes, I think your changes as they are written are backwards incompatible, as `make release` does not currently build the base images. Please find a transition plan -- we can edit the jobs to call some more broad `make` target that you reconfigure in your PR, or something.

kargakis commented:
> as make release does not currently build the base images

`make release` is self-contained, why is this an issue?

stevekuznetsov commented:
Today if a job calls `make release` we can't just let it keep doing that in the future -- your changes bring base image builds into the new `make release`.

stevekuznetsov commented:
We need to make a change to the jobs such that 

 a. no changes are necessary for this PR to test correctly
 b. no changes are necessary once this PR merges for everything else to test correctly

kargakis commented:
> Today if a job calls make release we can't just let it keep doing that in the future -- your changes bring base image builds into the new make release.

Do we have any jobs that are not part of the sjb branch apart from the fork_ami job? A quick search shows that wherever we do `make release` we also call `hack/build-base-images.sh`.

stevekuznetsov commented:
> Do we have any jobs that are not part of the sjb branch

I think we've moved all of the Origin jobs into `sjb`.

kargakis commented on images/dockerregistry/Dockerfile:
```diff
@@ -6,20 +6,21 @@
- && rpm -V ${INSTALL_PKGS} \
- && yum clean all
-
 COPY config.yml ${REGISTRY_CONFIGURATION_PATH}
```
So this moved because that's how it is in dist-git. I didn't bother updating those images.

smarterclayton commented:
[test]

smarterclayton commented on hack/build-base-images.sh:
```diff
@@ -8,7 +8,15 @@ source "$(dirname "${BASH_SOURCE}")/lib/init.sh"
+os::util::ensure::gopath_binary_exists imagebuilder
+
+# image builds require RPMs to have been built
+os::build::release::check_for_rpms
```
I feel like the moment we do this the "base, release, images" separation doesn't make sense.  It's really "release, images".  Base is only supposed to be images that don't depend on build artifacts.

Would recommend moving this code out of here and into build-images (building images/source).  Put the check for rpms there (and remove the other hacks).

kargakis replied:
@smarterclayton @stevekuznetsov removed the RPM dependency from the base images and verified once more that `OS_BUILD_ENV_PRESERVE="_output/local/releases:_output/local/bin" hack/env make release` works. I'll squash, rebase, and wait for https://github.com/openshift/aos-cd-jobs/pull/221 before I start Jenkins on this one. In the meantime, ptal

kargakis commented:
Tests will not pass before we merge
https://github.com/openshift/aos-cd-jobs/pull/221/

On Thu, Apr 27, 2017 at 5:06 PM, OpenShift Bot <notifications@github.com>
wrote:

> continuous-integration/openshift-jenkins/test Running (
> https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin/990/)
> (Base Commit: decce00
> <https://github.com/openshift/origin/commit/decce00f97f117f2b2fe59d37a333207bb013d00>
> )
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/13642#issuecomment-297742319>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf6s-AmWCEe2smICLAEdsxVyWv5_vks5r0K7hgaJpZM4M0etR>
> .
>

kargakis commented:
[test]

kargakis commented:
[test]

kargakis commented:
https://github.com/openshift/origin/issues/13985 [test]

kargakis commented on images/dockerregistry/Dockerfile:
```diff
@@ -6,16 +6,22 @@
+ENV REGISTRY_CONFIGURATION_PATH=/config.yml \
+	DOCKER_REGISTRY_URL=${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT}
 
-CMD DOCKER_REGISTRY_URL=${DOCKER_REGISTRY_SERVICE_HOST}:${DOCKER_REGISTRY_SERVICE_PORT} /dockerregistry ${REGISTRY_CONFIGURATION_PATH}
```
@mfojtik @legionus for some reason when CMD is used, `DOCKER_REGISTRY_URL` is empty (not sure where exactly the defaulting is happening and we don't fail) but once I moved it to be an ENV, `DOCKER_REGISTRY_URL` is set to `:`. Inside the registry we just [validate that the env is non-empty](https://github.com/openshift/origin/blob/59948c60f19e0687471268828bbc309d8051fe98/pkg/dockerregistry/server/repositorymiddleware.go#L169) and eventually a registry with such a configuration fails to accept pushes (https://pastebin.com/if1RAZxa). I tried to track down why or where the defaulting is happening in case we use CMD but I failed. Any ideas?

kargakis replied:
@legionus can you help me understand why the registry works when this `DOCKER_REGISTRY_URL` env is not set? It seems like we use it only for testing.

kargakis commented:
```
/usr/lib/golang/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status

# github.com/openshift/origin/cmd/openshift
/usr/lib/golang/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status
```
This specific error is hit in `test_pull_request_origin_extended_networking_minimal` and `test_pull_request_origin_integration`. Both of these jobs are based on  `common/test_cases/origin_release.yml` which calls `make release` that actually breaks. `common/test_cases/origin_release.yml` is based on `common/test_cases/origin.yml` which is based on the `base` ami. The base ami should have all dependencies installed - not sure why the failure above.

@stevekuznetsov got any idea ?

stevekuznetsov commented:
Hmm.... not sure -- /cc @smarterclayton 

The workaround will make the jobs functional but I am concerned that we break `make release` with this patch.

stevekuznetsov commented:
@kargakis @smarterclayton what is the status here? Having a rpm-->image flow will significantly change how we are doing to be doing the vendoring stuff

kargakis commented:
> @kargakis @smarterclayton what is the status here? Having a rpm-->image flow will significantly change how we are doing to be doing the vendoring stuff

This is blocked on https://github.com/openshift/aos-cd-jobs/pull/237 AFAIK

kargakis commented:
[test]

kargakis commented:
@smarterclayton @stevekuznetsov all comments addressed, https://github.com/openshift/aos-cd-jobs/pull/237 is merged and tests are green.

stevekuznetsov commented:
[merge]

stevekuznetsov commented:
@kargakis @smarterclayton man ... merge tag --> first in queue --> no flakes ... it's too good. I'm suspicious lol

smarterclayton commented:
After this change the size of the docker images created went up pretty massively:

```
openshift/origin-docker-registry                           v3.6.0-alpha.1                             daf507c7fb50        8 weeks ago         429 MB
openshift/origin-haproxy-router                            v3.6.0-alpha.1                             fcdf6f3daf3d        8 weeks ago         656 MB
openshift/origin                                           v3.6.0-alpha.1                             4f46cd210abb        8 weeks ago         635 MB
1
```

vs

```
openshift/origin-docker-registry                           v3.6.0-alpha.2                             4cef0c113a58        4 hours ago         1.1 GB
openshift/origin-haproxy-router                            v3.6.0-alpha.2                             832708e54105        4 hours ago         1.02 GB
openshift/origin                                           v3.6.0-alpha.2                             5b2cc7070378        4 hours ago         1 GB
```

I'd like to know why.

smarterclayton commented:
We probably can't ship with a 40% increase in total image size without at least understanding why it happened.

kargakis commented:
Also origin-pod
```
openshift/origin-pod                                    v3.6.0-alpha.2                454d0744b530        8 hours ago         213 MB
openshift/origin-pod                                    v3.6.0-alpha.1                853986e79d22        8 weeks ago         1.14 MB
```
The difference here is that before it was using scratch, now it is using origin/base. Do these images run once per node or once per pod?

kargakis commented:
So it seems that the additional ~300MB that exist between building openshift/origin with yum and copying the binaries is that yum thinks origin-clients is a dependency to the origin rpm thus it pulls that rpm in too. The router image is based on origin, hence its own size increase.  Yum in the dockerregistry image pulls both origin and origin-clients rpms.

kargakis commented:
Yeah, so yum is not hallucinating, all those deps are set in our spec file :)

stevekuznetsov commented:
`origin` is the top-level package, so I would expect the clients to be in there. I guess I also would have expected it to be there in the past, too. Do we have fewer symlinks? Did we see `openshift3/ose-` images increase in size?

kargakis commented:
> origin is the top-level package, so I would expect the clients to be in there. I guess I also would have expected it to be there in the past, too. 

Here's the spec requirement: https://github.com/openshift/origin/blob/cc2ed8f6223b48fd604799af243d5c8d6a8046d8/origin.spec#L74. It's possible I am parsing the whole thing wrong.

> Do we have fewer symlinks? 

Yes, prior to yum install, the origin image was just a copy of the binary and symlinking everything.

> Did we see openshift3/ose- images increase in size?

OSE images have been installing RPMs for quite some time before we switched Origin. I am not sure of their state prior to RPM installs.

stevekuznetsov commented:
Oh, so previously we did not have `oc`, we just symlinked `openshift cli`? Not sure that is technically a bug. Especially if it's weight we've been eating in the `openshift3/ose-` space since forever. /cc @smarterclayton

smarterclayton commented:
It's technically correct, and also really annoying.  I think openshift
"provides" origin clients already.

On Fri, Jun 9, 2017 at 10:10 AM, Steve Kuznetsov <notifications@github.com>
wrote:

> Oh, so previously we did not have oc, we just symlinked openshift cli?
> Not sure that is technically a bug. Especially if it's weight we've been
> eating in the openshift3/ose- space since forever. /cc @smarterclayton
> <https://github.com/smarterclayton>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/13642#issuecomment-307399329>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p5oBzaz_NZlQXCbWoufhTm3MZD0Vks5sCVJSgaJpZM4M0etR>
> .
>
